internal/stores/objects.go

```diff
@@ -259,3 +276,276 @@ func NewJSONObjectStore(dir string) (*JSONObjectStore, error) {
+		// TODO: once we migrate the contract store over to a relational
+		// db as well, we might want to put the contract ID here and
+		// have a contract table with a mapping of contract ID to host.
+		// That makes for better migrations.
+		Host []byte `gorm:"index; NOT NULL"`
+	}
+)
+
+// TableName implements the gorm.Tabler interface.
+func (dbObject) TableName() string { return "objects" }
+
+// TableName implements the gorm.Tabler interface.
+func (dbSlice) TableName() string { return "slices" }
+
+// TableName implements the gorm.Tabler interface.
+func (dbSlab) TableName() string { return "slabs" }
+
+// TableName implements the gorm.Tabler interface.
+func (dbShard) TableName() string { return "shards" }
+
+// TableName implements the gorm.Tabler interface.
+func (dbSector) TableName() string { return "sectors" }
+
+// NewSQLObjectStore creates a new SQLObjectStore connected to a DB through
+// conn.
+func NewSQLObjectStore(conn gorm.Dialector, migrate bool) (*SQLObjectStore, error) {
+	db, err := gorm.Open(conn, &gorm.Config{})
+	if err != nil {
+		return nil, err
+	}
+
+	if migrate {
+		// Create the tables.
+		tables := []interface{}{
+			&dbObject{},
+			&dbSlice{},
+			&dbSlab{},
+			&dbShard{},
+			&dbSector{},
+		}
+		if err := db.AutoMigrate(tables...); err != nil {
+			return nil, err
+		}
+		if res := db.Exec("PRAGMA foreign_keys = ON", nil); res.Error != nil {
+			return nil, res.Error
+		}
+	}
+
+	return &SQLObjectStore{
+		db: db,
+	}, nil
+}
+
+// Object turns a dbObject into a object.Object.
+func (o dbObject) Object() (object.Object, error) {
+	var objKey object.EncryptionKey
+	if err := objKey.UnmarshalText([]byte(o.Key)); err != nil {
+		return object.Object{}, err
+	}
+	obj := object.Object{
+		Key:   objKey,
+		Slabs: make([]slab.Slice, len(o.Slabs)),
+	}
+	for i, sl := range o.Slabs {
+		var slabKey slab.EncryptionKey
+		if err := slabKey.UnmarshalText([]byte(sl.Slab.Key)); err != nil {
+			return object.Object{}, err
+		}
+		obj.Slabs[i] = slab.Slice{
+			Slab: slab.Slab{
+				Key:       slabKey,
+				MinShards: sl.Slab.MinShards,
+				Shards:    make([]slab.Sector, len(sl.Slab.Shards)),
+			},
+			Offset: sl.Offset,
+			Length: sl.Length,
+		}
+		for j, sh := range sl.Slab.Shards {
+			copy(obj.Slabs[i].Shards[j].Host[:], sh.Sector.Host)
+			copy(obj.Slabs[i].Shards[j].Root[:], sh.Sector.Root)
+		}
+	}
+	return obj, nil
+}
+
+// List implements the bus.ObjectStore interface.
+func (s *SQLObjectStore) List(path string) ([]string, error) {
+	if !strings.HasSuffix(path, "/") {
+		panic("path must end in /")
+	}
+
+	inner := s.db.Model(&dbObject{}).Select("SUBSTR(id, ?) AS trimmed", len(path)+1).
+		Where("id LIKE ?", path+"%")
+	middle := s.db.Table("(?)", inner).
+		Select("trimmed, INSTR(trimmed, ?) AS slashindex", "/")
+	outer := s.db.Table("(?)", middle).
+		Select("CASE slashindex WHEN 0 THEN ? || trimmed ELSE ? || substr(trimmed, 0, slashindex+1) END AS result", path, path).
```

Comment:
I think this query warrants some documentation because it's completely unreadable if you don't know the syntax. I would also multiline `CASE` statements and explain `||` is an operator that concatenates.

```
CASE slashindex 
    WHEN 0 THEN ? || trimmed 
    ELSE ? || substr(trimmed, 0, slashindex+1) 
END AS result"
```